[scroll-animations] Add parsing support for timeline-scope

nmoucht opened the pull request:
#### 9cdb5c7eeb7babecc1ad09adc8ded30a5ebd772e
<pre>
[scroll-animations] Add parsing support for timeline-scope
<a href="https://bugs.webkit.org/show_bug.cgi?id=279634">https://bugs.webkit.org/show_bug.cgi?id=279634</a>
<a href="https://rdar.apple.com/135919980">rdar://135919980</a>

Reviewed by Tim Nguyen.

Add parsing support for timeline-scope property.

* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::TimelineScope::operator== const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTimelineScope):
(WebCore::RenderStyle::timelineScope const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setTimelineScope):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTimelineScope):

Canonical link: <a href="https://commits.webkit.org/284045@main">https://commits.webkit.org/284045@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6776468bb2022bff75e305f1fbe15fb1160164ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67197 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46576 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19829 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71232 "Built successfully") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69315 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18124 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53873 "Passed tests") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70264 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42827 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58144 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34407 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39499 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15537 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16684 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61464 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15878 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72935 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11156 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15227 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61351 "Found 3 new test failures: webanimations/accelerated-translate-animation-underlying-transform-changed-in-flight.html webanimations/accelerated-translate-animation-with-transform.html webanimations/accelerated-web-animation-with-single-interval-and-easing-y-axis-above-1.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11189 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61430 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9162 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2763 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10388 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42381 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43458 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44644 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -36,6 +36,24 @@ namespace WebCore {
 
 class WebAnimation;
 
+struct TimelineScope {
```
Maybe nest this within `AnimationTimeline`?

nmoucht replied:
I think generally we don't nest these structs used for parsing so we don't need to prefix them everywhere.

graouts replied:
There is precedent at least in the animation world. Take a look at `Animation.h` and notice `TransitionProperty` and `Direction` for instance. Not a big deal though.

graouts replied:
That said, if you're going to keep it outside the scope of `AnimationTimeline`, might be better to have it be its own `AnimationTimelineScope.h` file.

nmoucht replied:
Separated this to its own file.

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -36,6 +36,24 @@ namespace WebCore {
+        None,
+        All,
+        Ident
+    };
```
You can put all of those on a single line.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -4861,6 +4861,18 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+            return CSSPrimitiveValue::createCustomIdent(style.timelineScope().string);
+        default:
+            ASSERT_NOT_REACHED();
+            return CSSPrimitiveValue::create(CSSValueNone);
```
No need for the `default` since you've covered all three cases above. You can always add the `ASSERT_NOT_REACHED()` and `return` after the `switch()`.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2275,5 +2277,25 @@ inline RefPtr<TimingFunction> BuilderConverter::convertTimingFunction(BuilderSta
+            return { };
+        if (primitiveValue->valueID() == CSSValueAll)
+            return { TimelineScope::Type::All, { } };
+        return { TimelineScope::Type::Ident, { AtomString { primitiveValue->stringValue() } } };
```
use a switch ?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -4102,7 +4102,8 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         new DiscretePropertyWrapper<size_t>(CSSPropertyMaxLines, &RenderStyle::maxLines, &RenderStyle::setMaxLines),
-        new DiscretePropertyWrapper<OverflowContinue>(CSSPropertyContinue, &RenderStyle::overflowContinue, &RenderStyle::setOverflowContinue)
+        new DiscretePropertyWrapper<OverflowContinue>(CSSPropertyContinue, &RenderStyle::overflowContinue, &RenderStyle::setOverflowContinue),
+        new DiscretePropertyWrapper<const TimelineScope&>(CSSPropertyTimelineScope, &RenderStyle::timelineScope, &RenderStyle::setTimelineScope),
```
Spec says this is not animatable, so it should not have a wrapper: https://drafts.csswg.org/scroll-animations/#propdef-timeline-scope

nmoucht replied:
Fixed.

graouts commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-computed.tentative-expected.txt:
```diff
@@ -1,13 +1,13 @@
-FAIL The timeline-scope property shows up in CSSStyleDeclaration enumeration assert_not_equals: got disallowed value -1
-FAIL The timeline-scope property shows up in CSSStyleDeclaration.cssText assert_not_equals: got disallowed value -1
+PASS Property timeline-scope value 'initial'
+FAIL Property timeline-scope value 'inherit' assert_equals: expected "foo" but got "none"
```
Do you know why this is failing?

nmoucht replied:
I believe these tests are inaccurate now because timeline-scope was changed from custom-ident to dashed-ident (https://github.com/w3c/csswg-drafts/issues/8746). I'm planning to update them in a separate patch.

graouts commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-computed.tentative-expected.txt:
```diff
@@ -1,13 +1,13 @@
+PASS Property timeline-scope value 'unset'
+PASS Property timeline-scope value 'revert'
+PASS Property timeline-scope value 'none'
+FAIL Property timeline-scope value 'test' assert_true: 'test' is a supported value for timeline-scope. expected true got false
+FAIL Property timeline-scope value 'foo, bar' assert_true: 'foo, bar' is a supported value for timeline-scope. expected true got false
+FAIL Property timeline-scope value 'bar, foo' assert_true: 'bar, foo' is a supported value for timeline-scope. expected true got false
+FAIL Property timeline-scope value 'a, b, c, D, e' assert_true: 'a, b, c, D, e' is a supported value for timeline-scope. expected true got false
```
How about these failures?

nmoucht replied:
Same as above.

nmoucht committed:
[scroll-animations] Add parsing support for timeline-scope
https://bugs.webkit.org/show_bug.cgi?id=279634
rdar://135919980

Reviewed by Tim Nguyen.

Add parsing support for timeline-scope property.

* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::TimelineScope::operator== const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTimelineScope):
(WebCore::RenderStyle::timelineScope const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setTimelineScope):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTimelineScope):

Canonical link: https://commits.webkit.org/284045@main
